### PR TITLE
sstable address hashes type, sstables written to block storage without hierarchy

### LINKED ISSUE #1293: rocks: TierFS should store sstables directly under `committed.block_storage_prefix`
Per the design, sstables should be stored directly under `committed.block_storage_prefix` a-la `.git/objects` instead of the added hierarchy we currently have (with `$prefx/ranges` and `$prefix/meta-ranges`.

  #1201 was a prerequisite and is done.

### PR DESCRIPTION
**Pull request by ozkatz**
closes #1293

### CODE REVIEW

**Review by nopcoder — APPROVED** [review]

File: pyramid/tier_fs.go [hunk 1]
```diff
@@ -54,7 +54,7 @@ func NewFS(c *params.InstanceParams) (FS, error) {
 		fsLocalBaseDir: fsLocalBaseDir,
 		syncDir:        &directory{ceilingDir: fsLocalBaseDir},
 		keyLock:        cache.NewChanOnlyOne(),
-		remotePrefix:   path.Join(c.BlockStoragePrefix, c.FSName),
+		remotePrefix:   c.BlockStoragePrefix,
```
[the remark this answers is not in the document]
Comment (ozkatz) [reply pyramid/tier_fs.go:57]: Wasn't sure what else this was used for, but I vote to keep it for easier observability

File: ident/ident.go [hunk 2]
```diff
@@ -54,6 +54,10 @@ func NewAddressWriter() *AddressWriter {
 	return &AddressWriter{sha256.New()}
 }
 
+func WriterFromHash(h hash.Hash) *AddressWriter {
+	return &AddressWriter{h}
```
[the remark this answers is not in the document]
Comment (ozkatz) [reply ident/ident.go]: Agree. Added

File: graveler/sstable/writer.go [hunk 3]
```diff
@@ -7,6 +7,8 @@ import (
 	"hash"
 	"strconv"
 
+	"github.com/treeverse/lakefs/ident"
+
 	"github.com/treeverse/lakefs/graveler/committed"
```
[the remark this answers is not in the document]
Comment (ozkatz) [reply graveler/sstable/writer.go:12]: Done

File: pyramid/tier_fs.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ozkatz) [reply pyramid/tier_fs.go:57]: Also, it's still being used in the `fsLocalBaseDir` for writes, I think that's a good idea.

**Review by nopcoder — APPROVED** [review]

File: graveler/sstable/writer.go [hunk 4]
```diff
@@ -122,6 +123,10 @@ func (dw *DiskWriter) Abort() error {
 }
 
 func (dw *DiskWriter) Close() (*committed.WriteResult, error) {
+	// Before closing, we write all user supplied metadata keys and values to the hash
+	// This is done to avoid collisions, especially on empty sstables that might hash to the same value otherwise.
+	ident.WriterFromHash(dw.hash).MarshalStringMap(dw.props)
```
Comment (arielshaqed) [inline graveler/sstable/writer.go:128]: I don't understand why this drags `ident` in like this.  We already have `dw.hash`, it is a `Writer`.  If `ident.MarshalStringMap` were a function then I think you could just say
```suggestion
	ident.MarshalStringMap(dw.hash, dw.props)
```
I think that would make it easier for me to understand the line.

**Review by arielshaqed — APPROVED** [review]
Great, but 2 requests (approving for velocity; not necessary for me to approve these requests or even your ignoring them):
1. Document `AddMetadata` that it metadata is incorporated into the identity.
2. Inline: suggestion to avoid `WriterFromHash` and make it easier to read.

Also (because I cannot count):
3. This breaks all existing rocks repos, so please post on lakefs Slack #dev to let people know how to fix existing rocks repos.


THANKS!

**Review by itaiad200 — APPROVED** [review]
LGTM

File: graveler/sstable/writer.go (hunk 4 above)
Comment (ozkatz) [reply graveler/sstable/writer.go]: I agree, refactored ident to make this possible. PTAL

**Review by arielshaqed — APPROVED** [review]
Thanks!  Sorry for so much copy-pasta.  (BTW, it hints that maybe `AddressWriter` just isn't a thing.  But certainly *do not remove it in this PR*.)

**Comment by ozkatz** [comment]
> Thanks! Sorry for so much copy-pasta. (BTW, it hints that maybe `AddressWriter` just isn't a thing. But certainly _do not remove it in this PR_.)

Indeed, I left it there for compatibility only.

**Pull request merged**